src/genome.py

```diff
@@ -88,154 +97,260 @@ def __str__(self) -> str:
         removed from the set of active TEs.
-
         Returns a new ID for the transposable element.
         """
-        ...  # FIXME
-        return -1
+
+        if len(self.TE) != 0:
+            TE_ID = max(self.TE) + 1
+        else:
+            TE_ID = 1
+        self.TE[TE_ID] = length
+        self.active_TE.append(TE_ID)
+        if self.nucleotide[pos] in self.active_TE:
+            self.disable_te(self.nucleotide[pos])
+        self.nucleotide[pos:pos] = [TE_ID]*length
+        return TE_ID
+
 
     def copy_te(self, te: int, offset: int) -> int | None:
         """
         Copy a transposable element.
-
-        Copy the transposable element te to an offset from its current
+        Copy the transposable element te to an offset from its current_node
         location.
-
         The offset can be positive or negative; if positive the te is copied
         upwards and if negative it is copied downwards. If the offset moves
         the copy left of index 0 or right of the largest index, it should
         wrap around, since the genome is circular.
-
         If te is not active, return None (and do not copy it).
         """
-        ...  # FIXME
+        if te not in self.active_TE:
+            return None
+        pos = self.nucleotide.index(te)
+        res = (pos + offset) % len(self.nucleotide)
+        return self.insert_te(res, self.TE[te])
+        
+            
 
     def disable_te(self, te: int) -> None:
         """
         Disable a TE.
-
         If te is an active TE, then make it inactive. Inactive
         TEs are already inactive, so there is no need to do anything
         for those.
         """
-        ...  # FIXME
+        self.active_TE.remove(te)
+        for i in range(len(self)):
+            if self.nucleotide[i] == te:
+                for j in range(self.TE[te]):
+                    self.nucleotide[i+j] = 'x'
+                break
 
     def active_tes(self) -> list[int]:
         """Get the active TE IDs."""
-        ...  # FIXME
-        return []
+        return self.active_TE
 
     def __len__(self) -> int:
-        """Current length of the genome."""
-        ...  # FIXME
-        return 0
+        """current_node length of the genome."""
+        return len(self.nucleotide)
 
     def __str__(self) -> str:
         """
         Return a string representation of the genome.
-
         Create a string that represents the genome. By nature, it will be
         linear, but imagine that the last character is immidiatetly followed
         by the first.
-
         The genome should start at position 0. Locations with no TE should be
         represented with the character '-', active TEs with 'A', and disabled
         TEs with 'x'.
         """
-        return "FIXME"
+        nucleotide = ""
+        for x in self.nucleotide:
+            if isinstance(x, int):
+                nucleotide += "A"
+            else: 
+                nucleotide += str(x)
```

Comment:
Here, you could deal with inactivated TEs as well. If `x` is an integer and `x not in self.activei_TE` add an `'x'`. Otherwise, if `x` is an integer and it is in `self.active_TE` add an `'A'`.